### PR TITLE
Requestify implicit imports

### PR DESCRIPTION
**Pull request by hamishknight**
Add `ModuleImplicitImportsRequest`, which computes the modules that should be implicitly imported by each file of a given module by consulting the `ImplicitImportInfo` that it's created with. Use this request in import resolution to add all the necessary implicit imports. 

Along with https://github.com/apple/llvm-project/pull/1079, this allows us to remove `SourceFile::addImports`, ensuring that a file's imports are computed only once.

Note I'm not planning on merging this until after the 5.3 re-branch.

### CODE REVIEW

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/1079

@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/1079

@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility Debug

**Comment by CodaFi** [comment]
Debug failure seems like a consequence of Slava's patches to opaque type reconstruction. I wouldn't worry about it.

File: lib/Sema/ImportResolution.cpp [hunk 1]
```diff
@@ -264,24 +276,19 @@ class ImportResolver final : public DeclVisitor<ImportResolver> {
...
+  FrontendStatsTracer tracer(SF.getASTContext().Stats,
+                             "Import resolution");
   ImportResolver resolver(SF);
 
   // Resolve each import declaration.
   for (auto D : SF.getTopLevelDecls())
     resolver.visit(D);
 
-  SF.addImports(resolver.getFinishedImports());
+  SF.setImports(resolver.getFinishedImports());
```
Comment (CodaFi) [inline lib/Sema/ImportResolution.cpp:291]: This seems ripe for requestification next.

File: lib/Sema/ImportResolution.cpp [hunk 2]
```diff
@@ -264,24 +276,19 @@ class ImportResolver final : public DeclVisitor<ImportResolver> {
...
-  FrontendStatsTracer tracer(SF.getASTContext().Stats,
-                             "Import resolution");
-
-  // Make sure we skip adding the standard library imports if the
-  // source file is empty.
-  if (SF.ASTStage == SourceFile::ImportsResolved ||
-      SF.getTopLevelDecls().empty()) {
-    SF.ASTStage = SourceFile::ImportsResolved;
+  // If we've already performed import resolution, bail.
+  if (SF.ASTStage == SourceFile::ImportsResolved)
```
Comment (CodaFi) [inline lib/Sema/ImportResolution.cpp:280]: I've always wondered why these stage guards aren't assertions.

**Review by CodaFi — APPROVED** [review]
Nice cleanups here.

File: lib/Sema/ImportResolution.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/Sema/ImportResolution.cpp:291]: Yup, it's where I'm heading next :)

File: lib/Sema/ImportResolution.cpp (hunk 2 above)
Comment (hamishknight) [reply lib/Sema/ImportResolution.cpp:280]: Good idea! Fortunately this can all go away once import resolution is requestified.

Comment (hamishknight) [reply lib/Sema/ImportResolution.cpp:280]: Ah nevermind, the frontend can currently call `performImportResolution` before calling `performTypeChecking` (which also calls `performImportResolution`). I'll leave this as-is, and rip it out when requestifying import resolution.

**Comment by hamishknight** [comment]
Rebased to resolve a merge conflict.

https://github.com/apple/llvm-project/pull/1079

@swift-ci please test

**Comment by hamishknight** [comment]
Rebased to resolve a merge conflict.

https://github.com/apple/llvm-project/pull/1079

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/1079

@swift-ci please test Linux

**Pull request merged**